iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp

```diff
@@ -47,39 +47,33 @@ static Optional<Type> convertRank0TensorToScalar(RankedTensorType tensorType) {
   return elementType;
 }
 
-static Optional<Value> materializeCastToSignless(OpBuilder &builder,
-                                                 IntegerType toType,
-                                                 ValueRange inputs,
-                                                 Location loc) {
-  assert(inputs.size() == 1 && "too many inputs to type conversion");
-  Value fromValue = inputs[0];
-  auto fromType = fromValue.getType();
-  if (fromType.isSignlessInteger() || !toType.isSignlessInteger())
-    return llvm::None;
-  // Use unrealized conversion casts to do signful->signless conversion.
-  return builder.create<UnrealizedConversionCastOp>(loc, toType, fromValue)
-      ->getResult(0);
+static Type convertShapedToSignless(ShapedType shapedType) {
+  if (auto intType = shapedType.getElementType().dyn_cast<IntegerType>())
+    return shapedType.clone(convertIntegerToSignless(intType));
+  return shapedType;
 }
 
-static Optional<Value> materializeCastToScalar(OpBuilder &builder, Type toType,
-                                               ValueRange inputs,
-                                               Location loc) {
+static Optional<Value> materializeCast(OpBuilder &builder, Type toType,
+                                       ValueRange inputs, Location loc) {
   assert(inputs.size() == 1 && "too many inputs to type conversion");
   Value fromValue = inputs[0];
   auto fromType = fromValue.getType().dyn_cast<RankedTensorType>();
-  if (!fromType || fromType.getRank() != 0) return llvm::None;
+  if (!fromType) return llvm::None;
 
   if (auto intFromType = fromType.getElementType().dyn_cast<IntegerType>()) {
-    if (!intFromType.isSignlessInteger()) {
-      if (!toType.isSignlessInteger()) return llvm::None;
-      fromType = fromType.clone(toType).cast<RankedTensorType>();
+    Type castType = getElementTypeOrSelf(toType);
+    if (auto shapedType = fromType.dyn_cast<ShapedType>())
+      castType = shapedType.clone(castType);
+
+    if (castType != fromType)
       fromValue =
-          builder.create<UnrealizedConversionCastOp>(loc, fromType, fromValue)
+          builder.create<UnrealizedConversionCastOp>(loc, castType, fromValue)
```

Comment:
Still unclear why we need to introduce this ourselves. The flow with dialect converter that I am used to is that If there is an illegal operation, legalize it. If the operands types dont match up then the dialect conversion itself will introduce an unrealized conversion cast. IIUC, the unrealized conversion cast is a internal implementation detail of the dialect conversion framework.

Reply:
In this case we potentially need to insert two operations - a cast and an extract. If we relied on the pure UnrealizedConversionCast we would not get an extract op inserted.

Reply:
I think this function is doing too much at once. Separating them out might make this cleaner. Anyway, not holding this up on this anymore.

Reply:
I think the issue is that we're not doing a function conversion in this pass so the previous function signature doesn't get converted. That's why we need to explicitly cast in the sort op signature conversion. The types come in aren't converted, so the dialect conversion framework doesn't insert a cast. To me this points to the need to combine the MHLO conversion to linalg and to linalg extension. I also agree with Mahesh that this function is a bit overloaded now but not too big a deal. You can register multiple materialization functions I believe, so this could be split out.